src/lib/snarky_group_hash/snarky_unchecked_hash.ml

```diff
@@ -0,0 +1,342 @@
+    in
+    Sexp.of_string_conv_exn s' F.t_of_sexp
+
+  let field_map = List.map ~f:(fun (a, b) -> (a, field_of_string b))
+
+  let n1tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n1_tuples)
+
+  let d1tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d1_tuples)
+
+  let n2tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n2_tuples)
+
+  let d2tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d2_tuples)
+
+  let n3tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map n3_tuples)
+
+  let d31tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d31_tuples)
+
+  let d32tbl = Hashtbl.of_alist_exn (module Hash_key) (field_map d32_tuples)
+
+  let sum (xs : F.t list) =
+    List.fold ~init:F.zero ~f:(fun x y -> F.Infix.(x + y)) xs
+
+  let product (xs : F.t list) =
+    List.fold ~init:F.one ~f:(fun x y -> F.Infix.(x * y)) xs
+
+  let powers (v : F.t) (k : int) : F.t list =
+    let rec make_powers acc prev x =
+      if x = k then acc
+      else
+        let prev' = F.Infix.(prev * v) in
+        let acc' = prev' :: acc in
+        make_powers acc' prev' (x + 1)
+    in
+    List.rev (make_powers [F.one] F.one 0)
+
+  (* highest power of A is 24, of B is 16 *)
+  let a_powers = powers Params.a 24
+
+  let b_powers = powers Params.b 16
+
+  (*
+   * Xi(t) = Ni(t)/Di(t)  for i = 1, 2, 3
+   * N1(t) = A^2 . t . sum from j = 0 to j = 4 of [
+   *    sum for 2a+3b=3j of ( n1_(a,b) A^a . B^b ) . t^j ]
+   * D1(t) = sum from j = 0 to j = 5 of [
+   *    sum for 2a+3b=3j of ( d1_(a,b) A^a . B^b ) . t^j ]
+   *
+   * N2(t) = sum from j = 0 to j = 6 of [
+   *    sum for 2a+3b=3j of ( n2_(a,b) A^a . B^b ) . t^j ]
+   * D2(t) = 144At . sum from j = 0 to j = 4 of [
+   *    sum for 2a+3b=3j of ( d2_(a,b) A^a . B^b ) . t^j ]
+   *
+   * N3(t) = sum from j = 0 to j = 15 of [
+   *    sum for 2a+3b=3j+3 of ( n3_(a,b) A^a . B^b ) . t^j ]
+   * D3(t) = A . sum from j = 0 to j = 5 of [
+   *    sum for 2a+3b=3j of ( d31_(a,b) A^a . B^b ) . t^j ]
+   *    .
+   *    sum from j = 0 to j = 10 of [
+   *    sum for 2a+3b=3j of ( d32_(a,b) A^a . B^b ) . t^j ]
+   *)
+
+  let check x =
+    match x with
+    | Some x -> x
+    | None -> failwith "trying to mulitply by index not in powers lists"
+
+  let mul ~a ~b ~tbl =
+    F.Infix.(
+      Hashtbl.find_exn tbl (a, b)
+      * check (List.nth a_powers a)
+      * check (List.nth b_powers b))
+
+  let t_mul (sum : F.t) (t_power : int) (t_powers : F.t list) =
+    F.Infix.(sum * check (List.nth t_powers t_power))
+
+  let rec iter_a s a j acc tbl =
+    (* 2a + 3b = 3j -> b = 3j - 2a / 3 *)
+    (* 2a + 3b = 3j + 3 -> b = 3 + 3j - 2a / 3 *)
+    let b =
+      if s = 0 then ((3 * j) - (2 * a)) / 3 else (3 + (3 * j) - (2 * a)) / 3
+    in
+    let acc' = F.Infix.(acc + mul ~a ~b ~tbl) in
+    (* start with a_max and decrement until a = 0 *)
+    if a > 0 then
+      let a' = a - 1 in
+      iter_a s a' j acc' tbl
+    else acc'
+
+  let rec iter_j s j max acc t_powers tbl =
```

Comment:
Could you write the sums that `iter_j` and `iter_a` compute? I would also consider making the arguments labelled since a few of them have the same type.